galley/pkg/kube/source/init.go

```diff
@@ -15,18 +15,86 @@
 package source
 
 import (
+	"io/ioutil"
 	"fmt"
+	"os"
 	"time"
+	"strings"
 
+	"github.com/ghodss/yaml"
 	"github.com/hashicorp/go-multierror"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
 	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/util/wait"
-
 	"istio.io/istio/galley/pkg/kube"
 	kube_meta "istio.io/istio/galley/pkg/metadata/kube"
 )
 
+func writeCRDs(cs clientset.Interface, fileName string) (error) {
+	CRDData, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		scope.Infof("Read CRD %v", err)
+		return err
+	}
+
+	// yaml.YAMLtoJSON does not appear to understand "---".
+	yamlStrings := strings.Split(string(CRDData), "---")
+	// Build a set of yaml strings and range over them.
+	for _, yamlStringCRD := range yamlStrings {
+		var yamlCRD apiextensionsv1beta1.CustomResourceDefinition
+		yaml.Unmarshal([]byte(yamlStringCRD), &yamlCRD)
+
+		_, err = cs.ApiextensionsV1beta1().CustomResourceDefinitions().Create(&yamlCRD)
+		if err != nil {
+			scope.Infof("Error writing CRDs: %v", err)
+			continue
+//			return err
+		}
+	}
+	return nil
+}
+
+func createResourceTypes(cs clientset.Interface, CRDDirectory string) (error) {
+	CRDDirectoryFile, err := os.Open(CRDDirectory)
```

Comment:
This probably should be a single file, instead of a directory.

Reply:
Ah, I noticed that we're creating multiple files in the folder.

What is the expected use model here? As we up the versions, do we keep adding new crds yaml files for each version's delta? Can't we have a single file?

With future versioning support for CRDs, we will need to do in-place upgrade of CRDs to specify metadata about multiple versions.

Reply:
@ozevren the general thinking of the environments WG is that we offer a separate configmap per release.  In place upgrades of CRDs should be viable once we get out from under the weight of Helm proper (i.e. when we have a process managing the CRDs).  However, I'd rather burn that bridge when we get to it rather than speculate about implementation that I can't know in advance. :)

Reply:
If I am reading the code right, the current code would slurp all of the files and try to apply them together. It won't try to do version management.

As such the underlying premise of multiple files seems add-only updates from version to version. I think that assumption is wrong. The CRDs need to be in-place upgraded with a check to avoid overwriting a newer CRD (the assumption should be, newer CRDs are backwards compatible).